spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/query/SpannerStatementQueryTests.java

```diff
@@ -0,0 +1,268 @@
+
+		when(queryMethod.getName()).thenReturn(
+				"findTop3DistinctByActionAndSymbolOrTraderIdAndPriceLessThanOrPriceGreater"
+						+ "ThanEqualAndIdIsNotNullAndTraderIdIsNullAndTraderIdLikeAndPriceTrueAndPriceFalse"
+						+ "AndPriceGreaterThanAndPriceLessThanEqualOrderByIdDesc");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[]{
+		"BUY",
+		"abcd",
+		"abc123",
+		8.88,
+		3.33,
+		"ignored",
+		"ignored",
+		"blahblah",
+		"ignored",
+		"ignored",
+		1.11,
+		2.22,
+		};
+
+		when(spannerOperations.find(any(), (Statement) any(), any()))
+				.thenAnswer(invocation -> {
+					Statement statement = invocation.getArgument(1);
+					assertEquals(
+							"SELECT DISTINCT * FROM trades WHERE ( action=@tag0 AND ticker=@tag1 ) OR "
+									+ "( trader_id=@tag2 AND price<@tag3 ) OR ( price>=@tag4 AND id<>NULL AND "
+									+ "trader_id=NULL AND trader_id LIKE %@tag7 AND price=TRUE AND price=FALSE AND "
+									+ "price>@tag10 AND price<=@tag11 )ORDER BY id DESC LIMIT 3; {tag0: BUY, tag1:"
+									+ " abcd, tag8: ignored, tag9: ignored, tag10: 1.11, tag6: ignored, tag11: 2.22, "
+									+ "tag7: blahblah, tag4: 3.33, tag5: ignored, tag2: abc123, tag3: 8.88}",
+							statement.toString());
+					return null;
+				});
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void unspecifiedParametersTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn(
+				"findTop3DistinctIdActionPriceByActionAndSymbolOrTraderIdAndPriceLessThanOrPriceGreater"
+						+ "ThanEqualAndIdIsNotNullAndTraderIdIsNullOrderByIdDesc");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[] {
+				"BUY",
+				"abcd",
+				"abc123",
+		};
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void unsupportedParamTypeTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn(
+				"findTop3DistinctIdActionPriceByActionAndSymbolOrTraderIdAndPriceLessThanOrPriceGreater"
+						+ "ThanEqualAndIdIsNotNullAndTraderIdIsNullOrderByIdDesc");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[] {
+				"BUY",
+				"abcd",
+				"abc123",
+				8.88,
+				3.33,
+				new Trade(),
+				"ignored",
+		};
```

Comment:
readability/semantics: it is unclear what the unsupported paramtype is, please restructure

Reply:
Could you elaborate? We can only give params as mysterious array of objects.

Reply:
what are we testing here?

Reply:
the "new Trade()" is not a param type that Spanner can use. 
the name "unsupportedParamTest" is talking about that Trade param.

Reply:
1.) Tests should be a living documentation. My point is that by simply reading the test it is hard to infer that that parameter is the cornerstone of the test. An explanatory comment would be helpful, or extracting it to a variable like `Object unsupportedParam = new Trade()`. 
2.) I tried it deleting it, 3 and 3.0 - we get IllegalArgumentException in the first two cases, so the test passes, while it fails only with the 3.0 (double). Maybe we should match on a specific message instead just the class?

Reply:
I'll add a comment indicating what is causing the expected exception. But of course unit tests are by no means mathematical proofs. If one were to get hacker-y minded I'm sure the vast majority of unit tests in existence could be triggered in unexpected ways. 

The intention here is to trigger a specific condition at a specific line and that is what this test does.  I think it would be obtuse to make a new exception type just for this one case. I also don't like the idea of matching against the exact text of the exception, since the punctuation, spacing, or actual words of the exception message is not important to the information of the message.

Reply:
Thank you for the comment, clarity is key in saving time for future developers and ourselves. 
I agree with new exception type being too much, and on exact matching as well, that's why I usually match on partial patterns only. Error messages _are_ important - they are one of the first things you see as a developer who starts to work on a project, and they influence usability - hence having tests around them can be useful. I'll work out these examples in the next couple of days.